### ISSUE TITLE
[Tune] Multiple `tune.with_parameters` calls crashes Ray

### ISSUE THREAD

**Issue by amogkam** [ray-project/ray#16609]
Running this code
```
import ray
from ray import tune

ray.init(num_cpus=2)

def train_fn(config, extra=None):
    tune.report(metric=1)


trainable = tune.with_parameters(train_fn, extra=1)
tune.run(trainable)

ray.shutdown()

ray.init(num_cpus=2)

def train_fn_2(config, extra1, extra2):
    tune.report(metric=1)

trainable = tune.with_parameters(train_fn_2, extra1=1, extra2=2)
tune.run(trainable)

ray.shutdown()
```

results in this error 
```
== Status ==
Memory usage on this node: 10.9/16.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/2 CPUs, 0/0 GPUs, 0.0/4.3 GiB heap, 0.0/2.15 GiB objects
Result logdir: /Users/amog/ray_results/train_fn_2021-06-22_00-06-04
Number of trials: 1/1 (1 TERMINATED)
+----------------------+------------+-------+--------+------------------+----------+
| Trial name           | status     | loc   |   iter |   total time (s) |   metric |
|----------------------+------------+-------+--------+------------------+----------|
| train_fn_4f188_00000 | TERMINATED |       |      1 |      0.000719786 |        1 |
+----------------------+------------+-------+--------+------------------+----------+


2021-06-22 00:06:05,078	INFO tune.py:549 -- Total run time: 0.67 seconds (0.42 seconds for the tuning loop).
2021-06-22 00:06:06,256	WARNING worker.py:614 -- File descriptor limit 256 is too low for production servers and may result in connection errors. At least 8192 is recommended. --- Fix with 'ulimit -n 8192'
2021-06-22 00:06:07,335	INFO services.py:1267 -- View the Ray dashboard at http://127.0.0.1:8265
Abort trap: 6
```

It seems like there is state being carried over when using `tune.with_parameters` even though Ray is shutdown and started again in between the subsequent runs. The second run causes Ray to crash. Probably something to do with the fact that the Parameter Registry is being reused.

If I remove the `tune.with_parameters` this works fine.

**Comment by amogkam** [comment]
cc @richardliaw @krfricke

**Comment by amogkam** [comment]
A simple fix is to just re-initialize the Parameter Registry for each `tune.with_parameters` call which solves the issue. @richardliaw should I just make a PR for this, or do you have any thoughts on a better fix?

**Comment by krfricke** [comment]
So the problem here seems to be:

- `tune.with_parameters` stores object references in a dict
- This dict gets serialized and sent to workers
- Some automatic deserialization process fails when it's trying to access objects from an old session

I tried adding `parameter_registry.references.clear()` in the `with_parameters` function, and it fixes the problem for the above example. However, then this code fails:

```
import ray
from ray import tune

ray.init(num_cpus=2)

def train_fn(config, extra=None):
    print("train1", extra)
    tune.report(metric=1)

def train_fn_2(config, extra1, extra2):
    print("train2", extra1, extra2)
    tune.report(metric=1)


trainable = tune.with_parameters(train_fn, extra=1)
trainable2 = tune.with_parameters(train_fn_2, extra1=1, extra2=2)

tune.run(trainable, verbose=0)
tune.run(trainable2, verbose=0)

ray.shutdown()
```

Which makes sense, as the first trainable is serialized only after the second trainable cleared the registry.

So the solution here would be to not use a _global_ registry object, but a local one each time `tune.with_parameters()` is called. 

I created a PR in #16611 to solve this issue.
